File: queue/queue_ll.js
class Node {
    constructor(val) {
        this.val = val;
        this.prev = null;
        this.next = null;
    }
}

// This linked list is operating from right to left
// so the leftt most element is the first in element
// meaning the first element is at the head.
const a = new Node('a');
const b = new Node('b');
const c = new Node('c');
const d = new Node('d');


a.next = b;
b.prev = a;
b.next = c;
c.prev = b;
c.next = d;
d.prev = c;

// enqueue
function enqueue(node, tail) {
    tail.next = node;
    node.prev = tail;

    return node;
}

// dequeue
function dequeue(head) {
    head.next.prev = null;
    head.next = null;

    return head;
}

// get node
function findNode(nodeValue, head) {
    let current = head;

    while(current.next) {
        console.log(current.val);
        if(current.val === nodeValue) {
            return true;
        } else {
            current = current.next;
        }
    }

    return -1;
}

// get last
function getLastNode(node) {
    let current = node;
    while(current) {
        current = current.next;
        if(!current.next) {
            break;
        }
    }

    return current;
}

// peak
function peak(node) {
    let current = node;
    while(current) {
        current = current.prev;
        if(!current.prev) {
            break;
        }
    }

    return current;
}

let one = new Node(1);
console.log(enqueue(one, d))
console.log("Get last: ", getLastNode(a))
console.log(dequeue(a))
console.log("Get first node: ", peak(d))
console.log(findNode('c', b))

File: queue/queue_ll.py
class Node:
    def __init__(self, val) -> None:
        self.val = val
        self.prev = None
        self.next = None
    
    def __str__(self):
        prev = self.prev.val if self.prev else None
        next = self.next.val if self.next else None
        return f"value: {self.val}, prev: {prev}, next: {next}"

# a <-> b <-> c <-> d
a = Node('a')
b = Node('b')
c = Node('c')
d = Node('d')

a.next = b
b.prev = a
b.next = c
c.prev = b
c.next = d
d.prev = c

# This linked list is operating from left to right
# so the right most element is the first in element
# meaning the first element is at the tail.
def enqueue(node, head):
    head.prev = node
    node.next = node
    

    return node

def dequeue(tail):
    tail.prev.next = None
    tail.prev = None

    return tail

def getNode(nodeVal, tail):
    current = tail
    while current.val != nodeVal:
        current = current.prev
        if current.prev == None:
            return -1

    return nodeVal

def getLast(root):
    head = root
    while head.prev != None:
        head = head.prev
    
    return head

def peek(node):
    tail = node
    while tail.next != None:
        tail = tail.next
    
    return tail

one = Node(1)
enqueue(one, a)
print(getLast(d))
print(dequeue(d))
print(peek(a))